nvmeof: provisioner add features

gadididi opened the pull request:
# Describe what this PR does #

This PR introduces foundational improvements to the NVMe-oF CSI provisioner, focusing on metadata handling, operation ordering, and API consistency. 

## Is there anything that requires special attention ##

- This PR still in progress!!
- **Volume context key changes**: This modifies the CSI interface contract. Node plugin now expects clean keys like `subsystemNQN` instead of `.rbd.nvmeof.SubsystemNQN`
- **Idempotent operations**: Gateway operations now handle "already exists" scenarios as success rather than error

## Key Changes:
- **Metadata restructuring**: Moved `NVMeoFVolumeData` to dedicated `volume.go` file and added namespace UUID tracking
- **Volume context cleanup**: Removed `.rbd.nvmeof.` prefix from CSI volume context keys while maintaining prefixed keys for internal RBD metadata storage
- **Operation ordering fix**: Reordered gateway operations to create listeners before namespaces for proper resource dependency handling
- **Idempotent operations**: Made core gateway operations (`AddHost`, `CreateListener`, `CreateSubsystem`, `AddNamespace`) idempotent to handle retry scenarios gracefully


## Related issues ##

https://github.com/ceph/ceph-csi/issues/5370

## Future concerns (mentioned in issue above) ##

- Multiple listener support (noted in TODO comments)
- Comprehensive cleanup/rollback mechanisms for failed operations are implemented here https://github.com/ceph/ceph-csi/pull/5538

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

gadididi committed:
nvmeof: makes AddHost,createListener, CreateSubsystem and AddNamespace idempotent

dont return an error, due to k8s can retrigger CreateVolume with same request.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: remove common prefix "rbd.nvmeof." in volumeContext

remove "rbd.nvmeof." prefix

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: move NVMeoFVolumeData to volume.go

creates volume.go file. it will in charge of metadata stuff

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi commented:
Hi @yati1998 , @Rakshith-R , can you please review my PR, thanks!

Rakshith-R commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -304,13 +305,11 @@ func createNVMeoFResources(
 	log.DebugLog(ctx, "Namespace created: %s/%s with NSID: %d", rbdPoolName, rbdImageName, nsid)
 	nvmeofData.NamespaceID = nsid
 
-	// Step 5: Create listeners
```
Removal of listeners should be moved to previous commit

gadididi replied:
fixed

Rakshith-R commented on internal/nvmeof/nvmeof.go:
```diff
@@ -132,10 +132,29 @@ func (gw *GatewayRpcClient) CreateNamespace(ctx context.Context, subsystemNQN, p
-	}
-	if resp.GetStatus() != 0 {
+	case resp.GetStatus() == int32(syscall.EEXIST):
+		r, err := gw.client.ListNamespaces(ctx, &pb.ListNamespacesReq{Subsystem: subsystemNQN})
+		switch {
+		case err != nil:
+			return 0, fmt.Errorf("failed to get namespaces in subsystem %s: %w", subsystemNQN, err)
+		case r.GetStatus() != 0:
+			return 0, fmt.Errorf("could not get namespaces in subsystem %s (status=%d): %s",
+				subsystemNQN, r.GetStatus(), r.GetErrorMessage())
+		case len(r.GetNamespaces()) == 0:
+			return 0, fmt.Errorf("no existing namespaces in subsystem %s", subsystemNQN)
+		}
+
+		for _, ns := range r.GetNamespaces() {
+			if ns.GetRbdPoolName() == poolName && ns.GetRbdImageName() == imageName {
+				return ns.GetNsid(), nil
+			}
+		}
+
+		return 0, fmt.Errorf("could not find existing namespace for %s/%s in subsystem %s", poolName, imageName, subsystemNQN)
```
Can we maybe add a helper function for this code block ?

gadididi replied:
fixed

gadididi committed:
nvmeof: add ns-uuid property into NVMeoFVolumeData

added, because in node plugin it is in used.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

Rakshith-R commented on internal/nvmeof/nvmeof.go:
```diff
@@ -332,13 +360,18 @@ func (gw *GatewayRpcClient) CreateListener(ctx context.Context, subsystemNQN str
-		return fmt.Errorf("gateway AddListener returned error (status=%d): %s", resp.GetStatus(), resp.GetErrorMessage())
+	if resp.GetStatus() == 0 {
+		log.DebugLog(ctx, "Listener added successfully: %s to subsystem %s", listenerInfo.Address, subsystemNQN)
+
+		return nil
 	}
+	if resp.GetStatus() == int32(syscall.EEXIST) { // EEXIST
+		log.DebugLog(ctx, "Listener %s already created for subsystem %s", listenerInfo.Address, subsystemNQN)
 
-	log.DebugLog(ctx, "Listener added successfully: %s to subsystem %s", listenerInfo.Address, subsystemNQN)
+		return nil // Listener already created, no error
+	}
 
-	return nil
+	return fmt.Errorf("gateway AddListener returned error (status=%d): %s", resp.GetStatus(), resp.GetErrorMessage())
```
Let's handle status in the following order for clarity :
- EEXISTs
- error (!=0)
- success (==0)

gadididi replied:
fixed

gadididi committed:
nvmeof: change the order create listener before create ns

change the order: create listener to be located after create subsystem and before create ns.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

Rakshith-R commented on internal/nvmeof/nvmeof.go:
```diff
@@ -306,12 +311,18 @@ func (gw *GatewayRpcClient) AddHost(ctx context.Context, subsystemNQN, hostNQN s
+		log.DebugLog(ctx, "Host %s already added to subsystem %s", hostNQN, subsystemNQN)
 
-	return nil
+		return nil // Host already added, no error
```
Need a newline between return and log to satisfy linter.

Rakshith-R replied:
Oh there is a newline in 321, 
please ignore this

Rakshith-R reviewed (commented):
Can you please modify commit text from
`nvmeof: controlelr: remove prefix in volumeContext`
to
`nvmeof: remove common prefix "rbd.nvmeof." in volumeContext`
?

Rest looks good to me

gadididi commented:
@mergifyio requeue

gadididi commented:
@mergifyio requeue

Rakshith-R commented:
@mergifyio requeue
